kerastuner/tuners/bayesian.py

```diff
@@ -0,0 +1,285 @@
+import numpy as np
+import random
+import json
+from scipy import optimize as scipy_optimize
+from sklearn import gaussian_process
+
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
+from ..engine import tuner as tuner_module
+from ..engine import oracle as oracle_module
+from ..engine import hyperparameters as hp_module
+
+
+class BayesianOptimizationOracle(oracle_module.Oracle):
+    """Bayesian optimization oracle.
+
+    It uses Bayesian optimization with a underlying Gaussian process model.
+    The acquisition function used is upper confidence bound (UCB).
+
+    Attributes:
+        init_samples: Int. The number of randomly generated samples as initial
+            training data for Bayesian optimization.
+        alpha: Float. Value added to the diagonal of the kernel matrix
+            during fitting.
+        beta: Float. The balancing factor of exploration and exploitation.
+            The larger it is, the more explorative it is.
+        seed: Int. Random seed.
+
+    """
+
+    def __init__(self,
+                 init_samples=2,
+                 alpha=1e-4,
+                 beta=2.6,
+                 seed=None):
+        super(BayesianOptimizationOracle, self).__init__()
+        self.init_samples = init_samples
+        self.alpha = alpha
+        self.beta = beta
+        self.seed = seed or random.randint(1, 1e4)
+        self._seed_state = self.seed
+        self._tried_so_far = set()
+        self._max_collisions = 20
+        self._num_trials = 0
+        self._score = {}
+        self._values = {}
+        self._x = None
+        self._y = None
+        self.gpr = gaussian_process.GaussianProcessRegressor(
+            kernel=gaussian_process.kernels.ConstantKernel(1.0),
+            alpha=self.alpha)
+
+    def populate_space(self, trial_id, space):
+        self.update_space(space)
+        # Generate enough samples before training Gaussian process.
+        if self._num_trials < self.init_samples or len(self._score) < 2:
+            self._num_trials += 1
+            values = self._new_trial()
+            self._values[trial_id] = values
+            return {'status': 'RUN', 'values': values}
+
+        self._training_data()
```

Comment:
Would a more explicit approach be more readable here? e.g. `y, x = self._get_training_data(values); self.gpr.fit(x, y)`

Currently it's not clear from reading the flow what `self._training_data()` is supposed to do.